File: src/calc_gains.jl
using BenchmarkTools
using Distributed
using TimerOutputs
using Tullio
using LoopVectorization

X = rand(54, 581012)*100;

# function get_gains!(X, current_values, idxs, gains)
#     Threads.@threads for i in eachindex(idxs)
#         s = 0.0
#         for j in eachindex(current_values)
#             @inbounds s += @fastmath sqrt(current_values[j] + X[j, idxs[i]])
#         end
#         gains[i] = s
#     end
# end;
function get_gains!(X, current_values, idxs, gains)
    @tturbo for i in eachindex(idxs)
        s = 0.0
        for j in eachindex(current_values)
            s += sqrt(current_values[j] + X[j, idxs[i]])
        end
        gains[i] = s
    end
end;

function calculate_gains!(X, gains, current_values, idxs, current_concave_value_sum)
    get_gains!(X, current_values, idxs, gains)
    
    gains .-= current_concave_value_sum
    return gains
end;

@btime begin
    d, n = size(X);
    idxs = collect(1:n);

    current_values = zeros(Float64, d)
    current_concave_values = sqrt.(current_values)
    current_concave_values_sum = sum(current_concave_values)

    gains = zeros(Float64, n);
    calculate_gains!(X, gains, current_values, idxs, current_concave_values_sum);
end

File: src/calc_gains.py
import numpy as np
import timeit
from numba import njit, prange

def get_gains(parallel, fastmath, cache):
    @njit(parallel=parallel, fastmath=fastmath, cache=cache)
    def get_gains_(x, y, idxs, gains):
        for i in prange(idxs.shape[0]):
            idx = idxs[i]
            gains[i] = np.sqrt(y + x[idx, :]).sum()

        return gains

    return get_gains_


if __name__ == "__main__":
    d = 54
    n = 581012
    x = np.abs(np.random.randn(n, d))*100
    y = np.zeros(d, dtype="float64")
    idxs = np.arange(n)
    gains = np.zeros(n, dtype="float64")

    get_gains_ = get_gains(True, True, False)
    
    t = timeit.timeit("get_gains_(x, y, idxs, gains)", number=1, globals=globals())
    print("Elapsed time: ", round(t, 7)*1000, "ms.")